numba_dppy/dpnp_glue/dpnpimpl.py

```diff
@@ -0,0 +1,169 @@
+        self.fn_name = fn_name
+        self.type_names = type_names
+        super(RetrieveDpnpFnPtr, self).__init__(sig, get_pointer)
+
+class _DPNP_EXTENSION:
+    def __init__(self, name):
+        dpnp_lowering.ensure_dpnp(name)
+
+    @classmethod
+    def dpctl_get_current_queue(cls):
+        ret_type  = types.voidptr
+        sig       = signature(ret_type)
+        return types.ExternalFunction("DPCTLQueueMgr_GetCurrentQueue", sig)
+
+    @classmethod
+    def dpctl_malloc_shared(cls):
+        ret_type  = types.voidptr
+        sig       = signature(ret_type, types.int64, types.voidptr)
+        return types.ExternalFunction("DPCTLmalloc_shared", sig)
+
+    @classmethod
+    def dpctl_queue_memcpy(cls):
+        ret_type  = types.void
+        sig       = signature(ret_type, types.voidptr, types.voidptr, types.voidptr, types.int64)
+        return types.ExternalFunction("DPCTLQueue_Memcpy", sig)
+
+    @classmethod
+    def dpctl_free_with_queue(cls):
+        ret_type  = types.void
+        sig       = signature(ret_type, types.voidptr, types.voidptr)
+        return types.ExternalFunction("DPCTLfree_with_queue", sig)
+
+
+    @classmethod
+    def dpnp_sum(cls, fn_name, type_names):
+        ret_type  = types.void
+        sig       = signature(ret_type, types.voidptr, types.voidptr, types.int64)
+        f_ptr     = get_dpnp_fptr(fn_name, type_names)
+        def get_pointer(obj):
+            return f_ptr
+        return types.ExternalFunctionPointer(sig, get_pointer=get_pointer)
+
+    @classmethod
+    def dpnp_eig(cls, fn_name, type_names):
+        ret_type  = types.void
+        sig       = signature(ret_type, types.voidptr, types.voidptr, types.voidptr, types.int64)
+        f_ptr     = get_dpnp_fptr(fn_name, type_names)
+        def get_pointer(obj):
+            return f_ptr
+        return types.ExternalFunctionPointer(sig, get_pointer=get_pointer)
+
+
+@overload(stubs.dpnp.sum)
+def dpnp_sum_impl(a):
+    dpnp_extension = _DPNP_EXTENSION("sum")
+
+    dpnp_sum = dpnp_extension.dpnp_sum("dpnp_sum", [a.dtype.name, "NONE"])
+
+    get_sycl_queue = dpnp_extension.dpctl_get_current_queue()
+    allocate_usm_shared = dpnp_extension.dpctl_malloc_shared()
+    copy_usm = dpnp_extension.dpctl_queue_memcpy()
+    free_usm = dpnp_extension.dpctl_free_with_queue()
+
+    def dpnp_sum_impl(a):
+        if a.size == 0:
+            raise ValueError("Passed Empty array")
+
+        sycl_queue = get_sycl_queue()
+        a_usm = allocate_usm_shared(a.size * a.itemsize, sycl_queue)
+        copy_usm(sycl_queue, a_usm, a.ctypes, a.size * a.itemsize)
+
+        out_usm = allocate_usm_shared(a.itemsize, sycl_queue)
+
+        dpnp_sum(a_usm, out_usm, a.size)
+
+        out = np.empty(1, dtype=a.dtype)
+        copy_usm(sycl_queue, out.ctypes, out_usm, out.size * out.itemsize)
+
+        free_usm(a_usm, sycl_queue)
+        free_usm(out_usm, sycl_queue)
+
+        return out[0]
+
+    return dpnp_sum_impl
+
+'''
```

Comment:
I am not a fan of having commented out code in the source. Please remove this code and bring it back once ready as a fresh PR.